### PR TITLE
FEATURE: new site setting for max logins per ip per hour/minute

### PR DESCRIPTION
**Pull request by techAPJ**
Meta topic: https://meta.discourse.org/t/admin-configuration-for-rate-limit/57686

@SamSaffron can you please review this PR?

### CODE REVIEW

**Comment by techAPJ** [comment]
@SamSaffron ping.

File: spec/controllers/session_controller_spec.rb [hunk 1]
```diff
@@ -659,6 +659,22 @@ def post_login
         end
       end
     end
+
+    context 'rate limited' do
+      it 'rate limits login' do
+        SiteSetting.max_logins_per_ip_per_hour = 2
+        RateLimiter.stubs(:disabled?).returns(false)
```
Comment (tgxworld) [inline spec/controllers/session_controller_spec.rb]: I think it is better that we avoid stubbing here. Maybe we can add a method into `RateLimiter` that will allow us to enable the rate limit when we want to?

File: spec/controllers/session_controller_spec.rb [hunk 2]
```diff
@@ -659,6 +659,22 @@ def post_login
...
+        RateLimiter.clear_all!
+
+        ActionDispatch::Request.any_instance.stubs(:remote_ip).returns('10.0.0.1')
```
Comment (tgxworld) [inline spec/controllers/session_controller_spec.rb]: Does the `remote_ip` change in the test env? It'll be good if we can avoid stubbing here.

File: spec/controllers/session_controller_spec.rb [hunk 3]
```diff
@@ -659,6 +659,22 @@ def post_login
...
+        3.times do
+          xhr :post, :create, login: user.username, password: 'myawesomepassword'
+        end
+        expect(response).not_to be_success
```
Comment (tgxworld) [inline spec/controllers/session_controller_spec.rb]: I think we need to be checking the response each time we make a `post` request.

**Comment by tgxworld** [comment]
@techAPJ  Looks good to me :+1: One small note, controller specs are going to be deprecated in Rails 5 so it is generally better to write new tests as integration specs.

File: spec/controllers/session_controller_spec.rb (hunk 1 above)
Comment (techAPJ) [inline spec/controllers/session_controller_spec.rb]: Okay, it's been used in many other places. I'll look into it and push a new commit for it.

File: spec/controllers/session_controller_spec.rb (hunk 2 above)
Comment (techAPJ) [inline spec/controllers/session_controller_spec.rb]: Seems like it does not. Removed the stubbing.

File: spec/controllers/session_controller_spec.rb (hunk 3 above)
Comment (techAPJ) [inline spec/controllers/session_controller_spec.rb]: Okay, done.

**Comment by techAPJ** [comment]
Thanks for the review, @tgxworld! ❤️

**Pull request merged**